python/tvm/relay/memory_alloc.py

```diff
@@ -0,0 +1,274 @@
+
+    def compute_storage(self, tensor_type):
+        dtype = TVMType(tensor_type.dtype)
+        shape = [int(sh) for sh in tensor_type.shape]
+        size = 1
+        for sh in shape:
+            size *= sh
+        size *= (dtype.bits * dtype.lanes + 7) // 8
+        return expr.const(size, dtype=self.compute_dtype)
+
+    def make_static_allocation(self, scope, tensor_type, i):
+        """Allocate a tensor with a statically known shape."""
+        shape = [int(sh) for sh in tensor_type.shape]
+        if len(shape) == 0:
+            shape = expr.const(np.array([]).astype(
+                self.compute_dtype), dtype=self.compute_dtype)
+        else:
+            shape = expr.const(np.array(shape), dtype=self.compute_dtype)
+        size = self.compute_storage(tensor_type)
+        alignment = self.compute_alignment(tensor_type.dtype)
+        dtype = tensor_type.dtype
+        sto = scope.let(f"storage_{i}", self.alloc_storage(
+            size, alignment, dtype))
+        # TODO(@jroesch): There is a bug with typing based on the constant shape.
+        tensor = self.alloc_tensor(sto, shape, dtype, tensor_type.shape)
+        return scope.let(f"tensor_{i}", tensor)
+
+    def visit_let(self, let):
+        scope = ScopeBuilder()
+
+        self.scopes.append(scope)
+        while isinstance(let, expr.Let):
+            new_val = self.visit(let.value)
+            scope.let(let.var, new_val)
+            let = let.body
+
+        new_body = self.visit(let)
+        scope.ret(new_body)
+        self.scopes.pop()
+
+        return scope.get()
+
+    def visit_call(self, call):
+        if is_primitive(call):
+            # Because we are in ANF we do not need to visit the arguments.
+            scope = self.current_scope()
+            new_args = [self.visit(arg) for arg in call.args]
+            ins = expr.Tuple(new_args)
+            ret_type = call.checked_type
+
+            if ret_type.is_dynamic():
+                assert isinstance(ret_type, ty.TensorType)
+                shape_func_ins = []
+                engine = compile_engine.get()
+                cfunc = engine.lower_shape_func(call.op, self.target_host)
+                input_states = cfunc.shape_func_param_states
+
+                is_inputs = []
+                for i, (arg, state) in enumerate(zip(new_args, input_states)):
+                    state = int(state)
+                    # Pass Shapes
+                    if state == 2:
+                        sh_of = self.visit(self.shape_of(arg))
+                        shape_func_ins.append(
+                            scope.let(f"in_shape_{i}", sh_of))
+                        is_inputs.append(0)
+                    # Pass Inputs
+                    elif state == 1:
+                        new_arg = self.visit(arg)
+                        shape_func_ins.append(
+                            scope.let(f"in_shape_{i}", new_arg))
+                        is_inputs.append(1)
+                    else:
```

Comment:
Could you leave a TODO here to handle the case when state == 3 indicating both shape and input data are needed? Currently, I don't think we have such case. So just leave a todo.